### PR TITLE
storage/engine: close engine before copying SSTs

### PR DESCRIPTION
**Pull request by tamird**

### CODE REVIEW

File: pkg/storage/engine/bench_test.go [hunk 1]
```diff
@@ -533,12 +535,18 @@ func runMVCCDeleteRange(emk engineMaker, valueBytes int, b *testing.B) {
...
 
+	locDirty := dir + "_dirty"
+
 	for i := 0; i < b.N; i++ {
-		locDirty := dir + "_dirty"
 		if err := os.RemoveAll(locDirty); err != nil {
 			b.Fatal(err)
 		}
 		if err := copyDir(dir, locDirty); err != nil {
+			if os.IsNotExist(err) {
```
Comment (tschottdorf) [inline pkg/storage/engine/bench_test.go]: Hmm, this looks creepy. Can't you close `eng` before getting here? I assume that's why you get errors here in the first place.

**Comment by petermattis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/storage/engine/bench_test.go, line 545 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16191#-KlBNsgz6RBaHz-wvsbp-r3-545:-KlOaYayWttIlqeHfapj:bpm8d07) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf867b05f6c87211cca7576867ab9811a998d7b8/pkg/storage/engine/bench_test.go#L545)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Hmm, this looks creepy. Can't you close `eng` before getting here? I assume that's why you get errors here in the first place.
</blockquote></details>

Ditto what @tschottdorf said. Seems likely that a background compaction on `eng` is what is causing problems. I think `s/defer eng.Close()/eng.Close()/g` would fix the problem.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16191)*

**Comment by tamird** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/storage/engine/bench_test.go, line 545 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16191#-KlBNsgz6RBaHz-wvsbp-r3-545:-KlPWTTo27HAQ5qLWq8i:bejltt7) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf867b05f6c87211cca7576867ab9811a998d7b8/pkg/storage/engine/bench_test.go#L545)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Ditto what @tschottdorf said. Seems likely that a background compaction on `eng` is what is causing problems. I think `s/defer eng.Close()/eng.Close()/g` would fix the problem.
</blockquote></details>

Hm, I thought I looked for that and determined that all engines were closed by the time this was called, but I was obviously wrong. Thanks for keeping me honest!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16191)*

**Comment by petermattis** [comment]
:lgtm:

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16191#-:-KlPfHgZFHdgFLAHudEj:bnfp4nl)*

**Pull request merged**